File: lovebrew/data/classes/ctr.py
import os
import shutil
import subprocess
from pathlib import Path

from .console import Console


class CTR(Console):

    ENV_PATH = Path(os.getenv("DEVKITARM"))

    TEXTURE_EXTENSIONS = [".png", ".jpg", ".jpeg"]
    FONT_EXTENSIONS = [".ttf", ".otf"]
    SOURCE_EXTENSIONS = [".lua", ".t3x", ".bcfnt"]

    COMMANDS = {
        "texture": "tex3ds {src} --format=rgba8888 -z auto -o {dst}",
        "font": "mkbcfnt {src} -o {dst}",
        "meta": "smdhtool --create '{name}' '{desc}' '{author}' {icon} {dst}.smdh",
        "binary": "3dsxtool {elf} {dst}.3dsx --smdh={dst}.smdh --romfs={romfs}"
    }

    def __init__(self, data):
        super().__init__(data)
        self.build()

    def _get_destination_path(self, filepath, ext):
        """
            Gets the destination path for @filepath, adding @ext to the end
        """

        destination = self.build_directory / filepath.with_suffix(ext)
        destination.parent.mkdir(parents=True, exist_ok=True)

        return destination

    def _convert_texture(self, filepath):
        destination = self._get_destination_path(filepath, ".t3x")

        command = CTR.COMMANDS["texture"].format(src=filepath,
                                                 dst=destination)

        self._run_command(command)

    def _convert_font(self, filepath):
        destination = self._get_destination_path(filepath, ".bcfnt")

        command = CTR.COMMANDS["font"].format(src=filepath,
                                              dst=destination)

        self._run_command(command)

    def _copy_file(self, filepath):
        destination = self._get_destination_path(filepath, filepath.suffix)

        try:
            shutil.copy2(filepath, destination)
        except Exception as error:
            print(error)

    def _finalize(self):
        command = CTR.COMMANDS["meta"].format(name=self.name,
                                              author=self.author,
                                              desc=self.description,
                                              icon=self.get_icon(),
                                              dst=self.output_directory / self.name)

        self._run_command(command)

        command = CTR.COMMANDS["binary"].format(elf=self.get_binary(),
                                                name=self.name,
                                                dst=self.output_directory / self.name,
                                                romfs=self.build_directory / self.source_directory.name)

        self._run_command(command)

    def build(self):
        super().build()

        for file in self.source_directory.rglob("*"):
            if file.suffix in CTR.TEXTURE_EXTENSIONS:
                self._convert_texture(file)
            elif file.suffix in CTR.FONT_EXTENSIONS:
                self._convert_font(file)
            elif file.suffix in CTR.SOURCE_EXTENSIONS:
                self._copy_file(file)

        self._finalize()

    def __str__(self):
        return "Nintendo 3DS"

File: source/classes/ctr.nim
import strutils, strformat, os

import ../prompts
import ../config/configfile

import console
export console

## Command line stuff to run
let tex_cmd = "tex3ds '$1' --format=rgba8888 -z auto --border -o '$2.t3x'"
let fnt_cmd = "mkbcfnt '$1' -o '$2.bcfnt'"

let meta_cmd = "smdhtool --create '$1' '$2' '$3' '$4' '$5.smdh'"
let bin_cmd  = "3dsxtool '$1' '$2.3dsx' --smdh='$2.smdh'"

## Applicable conversions n such
let textures = @[".png", ".jpg", ".jpeg"]
let fonts    = @[".ttf", ".otf"]

type CTR* = ref object of ConsoleBase

proc getName*(self : CTR) : string = "Nintendo 3DS"
proc getBinaryName*(self : CTR) : string = "3DS.elf"
proc getIconExtension*(self : CTR) : string = "png"
proc getExtension*(self : CTR) : string = "3dsx"
proc convertFiles(self : CTR, source : string) : bool

proc publish*(self : CTR, source : string) : bool =
    if not self.convertFiles(source):
        return false

    # If we're building in "raw" mode, don't create a 3dsx
    if Config.shouldOutputRawData():
        return true

    let properDescription = self.description & " • " & self.version
    let binaryPath        = getBuildDirectory() & "/LOVEPotion"

    try:
        # Meta Command
        runCommand(meta_cmd.format(self.name, properDescription, self.author, self.getIcon(), binaryPath))

        # Binary Command
        runCommand(bin_cmd.format(self.getBinaryPath(), binaryPath))
    except CatchableError as error:
        echo(error.msg)
        return false

    return self.packGameDirectory(getRomFSDirectory())


proc convertFiles(self : CTR, source : string) : bool =
    echo("Converting and copying files..")

    # Get our important directories
    let romFSDirectory = getRomFSDirectory()

    # Ensure the required ELF binary exists. If it doesn't, we should abort and inform the user.
    # This should only be an issue if compiling non-raw builds
    let elfBinaryPath = self.getBinaryPath()

    if not elfBinaryPath.fileExists() and not Config.shouldOutputRawData():
        BUILD_FAIL.showFormatted(source, self.getName(), self.getBinaryName(), self.getBinaryPath())
        return

    # Walk through the source directory
    var hasError = false

    for path in walkDirRec(source, relative = true):
        let (dir, name, extension) = splitFile(path)

        let relativePath = fmt("{source}/{path}")
        let destination = fmt("{romFSDirectory}/{dir}")

        if isHidden(relativePath):
            continue

        # May or may not be horribly inefficient because it attempt to create all directories
        # for pretty much every file multiple times, but it shouldn't be *that* bad
        try:
            createDir(destination)

            # If the file we're currently looking at has its extension in one of these sets, operate on it
            if extension in textures:
                runCommand(tex_cmd.format(relativePath, fmt("{destination}/{name}")))
            elif extension in fonts:
                runCommand(fnt_cmd.format(relativePath, fmt("{destination}/{name}")))
            else:
                copyFile(relativePath, fmt("{destination}/{name}{extension}"))
        except CatchableError as error:
            echo(error.msg)
            hasError = true
            break

    if hasError:
        return false

    return true
